### PR TITLE
Avoid purging Revisions in use by third-party packages #10678

### PR DESCRIPTION
**Pull request by MeghanaNalla**
…party packages.

In this corrected version, I've also added the import from django.db.utils import ProtectedError and passed the stdout parameter to the purge_revisions function to handle the output properly.

Fixes #...

1) I have  wrapped the call to revision.delete() in the purge_revisions command with try/except and make it ignore ProtectedError also created a counter for counting every time a ProtectedError occurs and display this counter when the command finishes.

_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [X] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [X] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [X] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Comment by lb-** [comment]
Please update the description to have a link to.the issue being fixed. 

 #10678

**Comment by lb-** [comment]
Tests are failing with
> cannot import name 'ProtectedError' from 'django.db.utils'

Please ping when the CI needs to run again.

**Comment by lb-** [comment]
Starting CI again.

Note: you can ignore the 'backend' failure, this is currently broken, some improvements are on their way.

File: docs/reference/pages/model_reference.md [hunk 1]
```diff
@@ -728,11 +728,40 @@ Every time a page is edited, a new `Revision` is created and saved to the databa
...
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Stale revisions are revisions that are no longer associated with any live objects. These revisions can accumulate         over time as content is updated and new revisions are created. To manage database size and improve performance,           administrators can use the `purge_revisions` management command to remove stale revisions from the database.
+
+    Please exercise caution when running the `purge_revisions` command, as it will permanently delete stale revisions         that are no longer referenced by any live content. Make sure to back up important revisions or consider using a           ForeignKey with `on_delete=models.PROTECT` if there are specific revisions that you want to preserve and prevent from     being deleted.
+
+    Using ForeignKey with `on_delete=models.PROTECT`
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Wagtail implementors may want to use a `ForeignKey` field with `on_delete=models.PROTECT` when referring to a             specific revision they don't want to be deleted. By doing so, the revision will be protected from deletion as long as     it is referenced by any other live objects in the system.
```
Comment (laymonage) [inline docs/reference/pages/model_reference.md:747]: Please remove the extraneous whitespaces

```suggestion
Stale revisions are revisions that are no longer associated with any live objects. These revisions can accumulate over time as content is updated and new revisions are created. To manage database size and improve performance, administrators can use the `purge_revisions` management command to remove stale revisions from the database.

Please exercise caution when running the `purge_revisions` command, as it will permanently delete stale revisions that are no longer referenced by any live content. Make sure to back up important revisions or consider using a ForeignKey with `on_delete=models.PROTECT` if there are specific revisions that you want to preserve and prevent from     being deleted.

Using ForeignKey with `on_delete=models.PROTECT`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Wagtail implementors may want to use a `ForeignKey` field with `on_delete=models.PROTECT` when referring to a specific revision they don't want to be deleted. By doing so, the revision will be protected from deletion as long as it is referenced by any other live objects in the system.
```

Check the other paragraphs as well.

File: docs/reference/pages/model_reference.md [hunk 2]
```diff
@@ -728,11 +728,40 @@ Every time a page is edited, a new `Revision` is created and saved to the databa
...
+
+    Here's an example of using `on_delete=models.PROTECT` with a `ForeignKey` field in your model definition:
+
+    .. code-block:: python
+
+        from django.db import models
+        from wagtail.core.models import Revision
```
Comment (laymonage) [inline docs/reference/pages/model_reference.md:754]:
```suggestion
        from wagtail.models import Revision
```

File: docs/reference/pages/model_reference.md [hunk 3]
```diff
@@ -728,11 +728,40 @@ Every time a page is edited, a new `Revision` is created and saved to the databa
...
+
+        class MyModel(models.Model):
+            revision = models.ForeignKey(
+                Revision,
+                on_delete=models.PROTECT,
+                # Add other fields for your model
```
Comment (laymonage) [inline docs/reference/pages/model_reference.md:760]:
```suggestion
```

File: wagtail/management/commands/purge_revisions.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 from django.conf import settings
 from django.core.management.base import BaseCommand
 from django.db.models import Q
+import sys
```
Comment (laymonage) [inline wagtail/management/commands/purge_revisions.py:4]:
```suggestion
```

File: wagtail/management/commands/purge_revisions.py [hunk 5]
```diff
@@ -71,11 +72,23 @@ def purge_revisions(days=None, pages=True, non_pages=True):
...
+                deleted_revisions_count += 1
+            except  ProtectedError:
+                protected_error_count += 1
+                pass
 
+    if stdout:
+        stdout.write(
+            stdout.style.SUCCESS(
+                f"Deleted {deleted_revisions_count} revisions. "
+                f"Ignored {protected_error_count} revisions due to ProtectedError."
```
Comment (laymonage) [inline wagtail/management/commands/purge_revisions.py:91]:
```suggestion
                f"Ignored {protected_error_count} protected revisions."
```

File: wagtail/management/commands/purge_revisions.py [hunk 6]
```diff
@@ -71,11 +72,23 @@ def purge_revisions(days=None, pages=True, non_pages=True):
...
 
     for revision in purgeable_revisions.iterator():
-        # don't delete the latest revision
         if not revision.is_latest_revision():
-            revision.delete()
-            deleted_revisions_count += 1
+            try:
+                revision.delete()
+                deleted_revisions_count += 1
+            except  ProtectedError:
```
Comment (laymonage) [inline wagtail/management/commands/purge_revisions.py:83]:
```suggestion
            except ProtectedError:
```

File: wagtail/tests/test_management_commands.py [hunk 7]
```diff
@@ -620,21 +620,26 @@ class TestPurgeRevisionsCommandForPages(TestCase):
...
-        return self.page
 
+        self.revision = Revision.objects.create(revision=1, page=self.page)
+        self.revision.user = self.user
+        self.revision.save()
+
+    def test_purge_revisions_protected_error(self):
+        # Run the purge_revisions command
+        with self.assertRaises(SystemExit):
+            self.call_command('purge_revisions', '--force')
```
Comment (laymonage) [inline wagtail/tests/test_management_commands.py:641]: The changes in this test file do not seem to be related to the implementation changes at all. There is no `--force` option for the command...

Instead, we should have a model that references the `Revision` model with a `ForeignKey` that has `on_delete=models.PROTECT`, create an instance of that model with a stale revision pointed by the foreign key, and try to run the command.

File: wagtail/management/commands/purge_revisions.py [hunk 8]
```diff
@@ -71,11 +72,23 @@ def purge_revisions(days=None, pages=True, non_pages=True):
...
+                protected_error_count += 1
+                pass
```
Comment (laymonage) [inline wagtail/management/commands/purge_revisions.py:85]:
```suggestion
```

**Review by laymonage — CHANGES_REQUESTED** [review]
Thanks for giving this a go! The changes in the command implementation seems to make sense, but the tests and documentation still need some work.

Also, no offence (just curious), did you use any AI help to write this PR? I saw some hints that led to this assumption:
- Reference of `wagtail.core.models` even though we moved all models in there to `wagtail.models` a while ago.
- Mention of there being a `--force` option to `purge_revisions` when there isn't any.

Not that it's wrong to use AI (even I use it sometimes), but we need to exercise caution when using such tools, and a human understanding of the codebase is very important when making PRs to open source projects 🙂

**Comment by lb-** [comment]
@MeghanaNalla did you mean to close this PR? Is everything ok?

**Comment by NXPY123** [comment]
@lb- is the issue #10678 still Open? There's another PR that references the issue but it seems to be on wagtail-nest.

**Comment by lb-** [comment]
@NXPY123 I'm not really sure what happened here, maybe ask on the issue itself. No need to mention me specifically, just a comment is fine. 

Those that are watching the issue may be able add info. As for this PR, it looks like it was started and then cancelled, either intentionally or accidentally (branch was deleted). It's possible it could just need a new PR with the same changes / feedback addressed though. 

We are yet to hear back from @MeghanaNalla

**Comment by Stormheg** [comment]
Hi @NXPY123, the issue is still open. You are welcome to continue work on it since we haven't heard from @MeghanaNalla.

Create a new PR with the requested changes from this PR addressed. Be sure to mention the issue number and that you addressed the changes requested in this PR.

If you could, can you take the commits from this PR as starting point? That way @MeghanaNalla is credited for their contribution. If thats not possible, no worries,  We will credit them manually.

**Comment by NXPY123** [comment]
Are you suggesting working on a rebase of this branch on main?

**Comment by Stormheg** [comment]
The replacement PR #10961 has now been merged into main. Since that is obviously based on work done by @MeghanaNalla I've credited them too. Congratulations on your contribution!
